### ISSUE TITLE
Feed widget props from action provider

### ISSUE THREAD

**Issue by AdbC99** [FredrikOseberg/react-chatbot-kit#98]
Hi, 

I have a server that does the message parsing and sends back a config object in the same API call. I do that method in the action provider because I want to handle different responses with different handlers. 

Each handler has a react component that needs to receive the config object and render its respective react component with that config object. 

Is this possible? I tried mapToState and it didn't seem to update the state when I changed the state in the action provider.

Hope you can help,
A

**Comment by AdbC99** [comment]
I'm just looking at:

interface IChatbotMessageProps {
    message: string;
    withAvatar?: boolean;
    loading?: boolean;
    messages: any[];
    delay?: number;
    id: number;
    setState?: React.Dispatch<React.SetStateAction<any>>;
    customComponents?: ICustomComponents;
    customStyles: {
        backgroundColor: string;
    };
}

Can I use setState to set the state of the widget from the actionProvider? It seems to exist here, but not in the options in the IMessage interfacer in the ActionsProvider.createChatBotMessage

**Comment by FredrikOseberg** [comment]
> I'm just looking at:
> 
> interface IChatbotMessageProps { message: string; withAvatar?: boolean; loading?: boolean; messages: any[]; delay?: number; id: number; setState?: React.Dispatch<React.SetStateAction>; customComponents?: ICustomComponents; customStyles: { backgroundColor: string; }; }
> 
> Can I use setState to set the state of the widget from the actionProvider? It seems to exist here, but not in the options in the IMessage interfacer in the ActionsProvider.createChatBotMessage

Yes you can, the state will be passed in it's entirity to the widget. The only problem is that the widget will share the state (if you render multiple widgets).

**Comment by AdbC99** [comment]
Thanks, I found mapStateToProps was the issue. I had tidied up my call to add to the state in a function:
```
key = "some string"

 (key, value) => {
    this.setState((prevState) => ({
        ...prevState,
        key: value,
    }));
}
```

And mapStateToProps did not like that, when I did the call using strings it worked:

```
this.setState((prevState) => ({
        ...prevState,
        "key as a string": value,
    }));
```

**Comment by FredrikOseberg** [comment]
> Thanks, I found mapStateToProps was the issue. I had tidied up my call to add to the state in a function:
> 
> ```
> const key = "some string"
> 
> const addToState = (key, value) => {
>     this.setState((prevState) => ({
>         ...prevState,
>         key: value,
>     }));
> }
> ```
> 
> And mapStateToProps did not like that, when I did the call using strings it worked:
> 
> ```
> this.setState((prevState) => ({
>         ...prevState,
>         "key as a string": value,
>     }));
> ```

Ah, yes that makes sense. Your example should also work if you wrap the key in brackets like this: 
```
> const addToState = (key, value) => {
>     this.setState((prevState) => ({
>         ...prevState,
>         [key]: value,
>     }));
> }
```

Happy new year!

**Comment by christianjeromeperlas-ernest** [comment]
how can I extend the ActionProvider class to another class to use many times the ActionProvider?
